src/backend/distributed/operations/auto_shard_split.c

```diff
@@ -0,0 +1,422 @@
+	bool isnull;
+	Datum resultDatum = SPI_getbinval(tuple, tupletable->tupdesc, 1, &isnull);
+
+	if (!isnull)
+	{
+		ereport(DEBUG4, errmsg("%s", GetShardSplitQuery(shardinfo, resultDatum,
+														shardSplitMode)->data));
+		ExecuteSplitBackgroundJob(jobId, shardinfo, resultDatum, shardSplitMode);
+		SPI_finish();
+		return 1;
+	}
+	else
+	{
+		ereport(LOG, errmsg("No Splitpoints for shard split"));
+		SPI_finish();
+		return 0;
+	}
+}
+
+
+/*
+ * citus_auto_shard_split_start finds shards whose colocation group has total size greater than threshold
+ * and from that extracts the shard with the maximum size . It isolates the shards if the shard has common
+ * tenant values otherwise it 2 way splits it on the basis of average hash value.
+ *
+ */
+Datum
+citus_auto_shard_split_start(PG_FUNCTION_ARGS)
+{
+	ErrorOnConcurrentOperation();
+	StringInfo query = makeStringInfo();
+
+	/* This query is written to group the shards on the basis of colocation id and shardminvalue and get the groups whose sum of shardsize
+	 * are greater than a threshold and than extract the shard in them which has the maximum size. So for that first pg_dist_shard and citus_shards are joined followed by the joining of pg_dist_node
+	 * and citus_shards and finally joined by the table obtained by the grouping of colocation id and shardminvalue and shardsize exceeding the threshold.*/
+
+	appendStringInfo(
+		query,
+		" SELECT cs.shardid,pd.shardminvalue,pd.shardmaxvalue,cs.shard_size,pn.nodeid, max_sizes.total_sum"
+		" FROM pg_catalog.pg_dist_shard pd JOIN pg_catalog.citus_shards cs ON pd.shardid = cs.shardid JOIN pg_catalog.pg_dist_node pn ON cs.nodename = pn.nodename AND cs.nodeport= pn.nodeport"
+		" JOIN"
+		" ( select shardid , max_size , total_sum from (SELECT distinct first_value(shardid) OVER w as shardid, sum(shard_size) OVER (PARTITION BY colocation_id, shardminvalue) as total_sum, max(shard_size) OVER w as max_size"
+		" FROM citus_shards cs JOIN pg_dist_shard ps USING(shardid)"
+		" WINDOW w AS (PARTITION BY colocation_id, shardminvalue ORDER BY shard_size DESC))as t)"
+		" AS max_sizes ON cs.shardid=max_sizes.shardid AND cs.shard_size = max_sizes.max_size AND pd.shardminvalue <> pd.shardmaxvalue AND pd.shardminvalue <> ''"
+		);
+
+	ereport(DEBUG4, errmsg("%s", query->data));
+	Oid shardTransferModeOid = PG_GETARG_OID(0);
+	Datum enumLabelDatum = DirectFunctionCall1(enum_out, shardTransferModeOid);
+	char *shardSplitMode = DatumGetCString(enumLabelDatum);
+	ereport(DEBUG4, errmsg("%s", shardSplitMode));
+	if (SPI_connect() != SPI_OK_CONNECT)
+	{
+		elog(ERROR, "SPI_connect to the query failed");
+	}
+	if (SPI_exec(query->data, 0) != SPI_OK_SELECT)
+	{
+		elog(ERROR, "SPI_exec for the execution failed");
+	}
+
+	SPITupleTable *tupletable = SPI_tuptable;
+	int rowCount = SPI_processed;
+	bool isnull;
+	int64 jobId = CreateBackgroundJob("Automatic Shard Split",
+									  "Split using SplitPoints List");
+	int64 count = 0;
```

Comment:
we try to avoid very generic names, so better to use scheduledSplitCount

Reply:
used scheduledSplitCount